### DISCUSSION TITLE
Support to set <html lang /> from metadata?

### DISCUSSION THREAD

**Discussion by ViliamKopecky** [vercel/next.js#49415] [App Router]
As far as I know you can only set `<html lang={lang} />` in root layout. It can be dynamic from `[lang]` route, which is problematic when you want to have a default locale.

In my opinion `lang` should be configurable same as other metadata as layout cannot always know before route loads data.

```ts
export async function generateMetadata() {
  return {
    lang: "cs",
    title: "Ahoj",
  }
}
```

This should result in something like

```html
<html lang="cs">
  <head>
    <title>Ahoj</title>
  </head>
  <body>
...
```

Do you plan tu support this? Or maybe is there a workaround similar to `useServerInsertedHTML`.

I would like to avoid unnecessary usage of `middleware.ts`.

**Comment by ryami333** [comment]
I would also like to add my support for this feature, and I would also like to request other `<html>` element attributes such as `dir`. Perhaps a generic API would work, something like:

```ts
export async function generateMetadata() {
  return {
    htmlAttributes: {
      lang: "ar-AE",
      dir: "rtl",
    },
    title: "Ahoj",
  }
}
```

---

Here's a summary of the _currently available_ options/workaround we have, to demonstrate why I think we need framework-level support for this.

## Pattern 1. - `lang` can be derived from params

### Solution/Workaround 1a.

We can use a `layout` file, and simply pass `params.lang` to the `<html>` element.

```tsx
// app/[lang]/layout.tsx

export const Layout = ({ params, children }) => {
  return (
    <html lang={params.lang}>
      <body>{children}
    </html>
  );
}
```

The problem: if I now navigate from eg. `/en/foo` to `/de/foo`, the `lang` attribute will not be updated client-side.

### Solution/Workaround 1b.

~~We can use a `template` file, because that re-renders upon navigation (as opposed to `layout` files, which do not).~~

EDIT: actually we can't because `template` doesn't receive `params`.

### Solution/Workaround 1c.

We can change the `lang` attribute on the client-side only, with an effect.

```tsx
// `app/[lang]/layout.tsx`

import { LangChangeHandler } from "./LangeChangeHandler";

export const Layout = ({ children }) => {
  return (
    <html>
      <LangChangeHandler />
      <body>{children}</body>
    </html>
  );
}
```
```tsx
// app/LangChangeHandler.tsx

"use client";
import { useParams } from "next/navigation";

export const LangChangeHandler = () => {
  const { lang } = useParams();
  
  useEffect(() => {
    document.documentElement.setAttribute("lang", lang);
  }, [lang];
  
  return null;
};
```

The problem: this works on the client-side only. Pages will not be server-side rendered with an appropriate `lang` tag on the `<html>` element.

### Solution/Workaround 1d.

We could do _both_ 1c and 1a (or 1b). This would ensure that the template is server-side-rendered with the correct `lang` on page-load, and that it is correctly updated upon navigation.

The problem: we now have two sources-of-truth for our document's `lang` attribute.

**Reply by j7i** [reply]
@ryami333 It is worth mentioning that by following pattern 1c we might run into problems with adding a custom `not-found.js` file.
Thus, according to my current understanding/experience, it would not be possible to catch the 404 errors from its [lang] boundaries.
Additionally, by placing the `not-found.js` file in the root `/app` directory, it currently requires a `layout.js` file on the same level; which then further requires that the `<html>` tag is placed in the very outer root layout.

**Reply by ryami333** [reply]
Yeah none of these workarounds are without compromise and I didn't intend to present them as solutions - the solution here would be for Next.js to introduce an API (like metadata) for changing these attributes on a per-page/directory basis, but they seem completely uninterested in this.

**Reply by yunusbayraktaroglu** [reply]
Another problem is that the solution recommended in the official Next.js documentation does not allow for maintaining a component in the HTML without it being unmounted between routes.

**Reply by gentlee** [reply]
I thing about this approach is that layout must be in `[lang]` folder (seems obvious but anyway).

**Reply by RowinVanAmsterdam** [reply]
> I thing about this approach is that layout must be in `[lang]` folder (seems obvious but anyway).

I tried this and while this does work, there is a huge con. As soon as the 'lang' param changes, the complete HTML tree gets re-rendered since your <html> tag is inside a dynamic path [lang]. So you always want to keep your rootlayout on root level of your application.

**Comment by ryami333** [comment]
## Pattern 2. - `lang`/`dir` must be asynchronously computed from params

Consider what happens when the attribute cannot be trivially derived from `params`. To demonstrate, I'll use the `dir` attribute here because it's a more obvious use-case than for `lang`, but the reasoning is still applicable to the `lang` attribute.

Consider a "blog", where each blog-post is fetched by its "`slug`" property. The blog post object contains a `dir` property, with a value of either `"rtl"` or `"ltr"`.

```ts
// app/[slug]/page.tsx
import { BlogMetadataHandler } from './BlogMetadataHandler';
import { getPost } from './getPost';

export const Page = async ({ params }) => {
  const post = await getPost(params.slug);
  
  return (
    <>
      <BlogMetadataHandler lang={post.lang} dir={post.dir} />
      <h1>{post.title}</h1>
      </p>{post.content}</p>
    </>
  );
}
```
```tsx
// app/[slug]/BlogMetadataHandler.tsx
"use client";

export const BlogMetadataHandler = ({ lang, dir }) => {
  useEffect(() => {
    document.documentElement.setAttribute("lang", lang);
    document.documentElement.setAttribute("dir", dir);
  });
  
  return null;
}
```

The problem: this works on the client-side only! The layout will not be SSR'ed with any `lang` or `dir` attribute on the document element. And as far as I can tell, this is the _only_ way to set these attributes for this pattern where the attributes are not trivially derived from params. The OP briefly mentions middleware, but I haven't been able to find any solution or workaround that would leverage middleware.

It's clear to me that this simply _must_ be either be added to the `generateMetadata` API or given a dedicated API called something like `generateDocumentAttributes`.

**Comment by ViliamKopecky** [comment]
@ryami333

> **Solution/Workaround 1a.**
> The problem: if I now navigate from eg. /en/foo to /de/foo, the lang attribute will not be updated client-side.

I think layouts do refresh (via RSC) when their parameters change. The difference is keeping internal state of the component by remounting the component with changing `key` like `<Template key={routeParam}><Page /></...>`. But the parameters should be always fresh.

Problem is supporting default `lang` `/[lang]/[slug]/page.tsx` (or even `/[lang]/[[...path]]/page.tsx`) when you would like default lang `en` so you could have `/contact` and `/de/kontakt` - here the solution is `middleware.ts`, but I would love if it could be supported without the need for middleware.

I like idea of the dedicated api `generateDocumentAttributes` so it would provide only the data that actually are needed for layout to rerender. Or even something like `generateLayoutExtraParameters` which would allow you to pass anything to parent layouts and it would refresh when these parameters change - like dependencies of `useEffect` etc.

**Reply by fikkatra** [reply]
@ViliamKopecky could you elaborate on this:

> so you could have /contact and /de/kontakt - here the solution is middleware.ts

Can you explain how you could use middleware so you can support both routes and still keep the `lang` attribute up to date? As far as I can see, you only have these options:
* Either you move the root layout from `app/layout.tsx` to `app/[lang]/layout.tsx`. This allows you to correctly set `<html lang="<current lang>">`. However, you loose the ability to display content from routes that do not contain the language (e.g. `/contact`). Your only option is to reroute to a lang-route, e.g. `/en/contact`, which may not always be desirable. 
* OR you keep the layout in the root, at `app/layout.tsx`. You set `<html lang="<default lang>">` and you override the language on a lower level in a client side component. But this means there will be a mismatch between the `lang` attribute and the language of the content from SSR (until the client side kicks in), thus negatively impacting SEO.

Is there an option I'm missing?

**Reply by ViliamKopecky** [reply]
Yes, the first option without root layout. :( it's not good.

**Comment by ViliamKopecky** [comment]
This is a demo what I think could be achieved without middleware. https://middleware-cache-three.vercel.app

**Reply by outruller** [reply]
Hey, @ViliamKopecky!

Did I get it right that the demoed solution _not_ using middleware and essentially SSG friendly?
If so, would you mind sharing a repo please or at least giving a couple of hints on how to achieve this behavior?

Thanks!

**Reply by ViliamKopecky** [reply]
No, it’s with middleware. I just hope it should be possible without it. 🥲

**Reply by 6mint** [reply]
Hey, @ViliamKopecky!
Could you please share the code for the demo?
It would certainly be helpful in the context of the discussion.
Thank you 🙂

**Reply by rkanani108** [reply]
Hey, @ViliamKopecky !
Could you please share the code for the demo?  🥇

**Reply by richardtaylordawson** [reply]
@ViliamKopecky Also hoping to see the code for this if possible :)

**Comment by jasperhartong** [comment]
Also looking for a solution for this.

As far as I can see, there's now no way to set the lang attribute serverside AND keep a certain component mounted while navigating between locales (e.g. have the `NextTopLoader` correctly indicate the loading state while navigating from `/en` to `/nl`). All because the lang needs to be set in the root layout and thus this root layout will always be fully remounted.

**Comment by richardtaylordawson** [comment]
I have this same issue and find it insane that nextjs doesn't have an answer for this 🤔

**Comment by Shwvi** [comment]
Same issue here. I need change html attr in children layout component

**Comment by tomferez** [comment]
It is crazy that there is still not a possible workaround for such an essential issue.

**Comment by jpmaga** [comment]
For the past couple of months the workaround i managed to come up with is to change the layout to this:

```tsx
export default function RootLayout({ children }: { children: React.ReactNode }) {
	return <>{children}</>
}
```

and then on page.tsx: 
```tsx
return (
	<html lang={data.lang ?? "en"}>
		<head>
			<title>{data.seo.title}</title>
			...
		</head>
		<body>
			...
		</body>
	</html>
)
```
I don't use `next/link` for navigation, so this might not work in that case.

**Reply by richardtaylordawson** [reply]
NextJS requires the base layout to contain the HTML tag with the lang attribute, so this won't work

**Reply by ryami333** [reply]
> NextJS requires the base layout to contain the HTML tag with the lang attribute 

No, that's not true. This works exactly as @jpmaga demonstrated above.

---

That said, it's still not a valid workaround, because the whole point of layouts is that they are persistent during navigation. If your layouts are distinct for each URL, then they don't have any utility in many cases and you might as well not have any `layout.tsx` and just set the `html` tags in your `page.tsx` files.

**Reply by richardtaylordawson** [reply]
I think that's fair. I think I read the following from the docs as if it was "necessary". But the language is open-ended 

`A root layout is the top-most layout in the root app directory. It is used to define the <html> and <body> tags and other globally shared UI.`

**Reply by jpmaga** [reply]
> NextJS requires the base layout to contain the HTML tag with the lang attribute, so this won't work

https://stackblitz.com/edit/nextjs-gm9kup

@ryami333:
> That said, it's still not a valid workaround, because the whole point of layouts is that they are persistent during navigation. If your layouts are distinct for each URL, then they don't have any utility in many cases and you might as well not have any `layout.tsx` and just set the `html` tags in your `page.tsx` files.

It is a workaround for my use case, in the sense that I don't use nextjs's navigation. I do however agree with you that it is not a proper workaround for all use cases. Ideally we'd have a solution to set it natively, using the generateMetada for instance. I would remove the layout.tsx if i could, but nextjs kindly creates it for you everytime you try to do that.

**Comment by yandodov** [comment]
There's a dirty workaround using a global state variable. For example, in `@/config.ts`:

```ts
export const state = { locale: "en" };
```

…then you update the variable in `app/[locale]/page.tsx`:

```tsx
import { state } from "@/config";

export default function Page({ params: { locale } }) {
	state.locale = locale;
	return (
		<div>
		...
		</div>
	);
}
```

…and use it in the root layout `app/layout.tsx`:

```tsx
import { state } from "@/config";

export default function RootLayout({ children }) {
	return (
		<html lang={state.locale}>
			<body>{children}</body>
		</html>
	);
}
```

This, of course, is a terrible pattern in general, but **it works**. At the end of the day, the staticly generated HTML has the correct `lang` and that's what matters.

You can find an explanation of this approach [in this article](https://dodov.dev/blog/how-to-set-up-simple-nextjs-localization) and you can also play around with a working version [in this repo](https://github.com/hdodov/nextjs-simple-localization).

**Reply by MonstraG** [reply]
Just did this, it worked. 

To be honest, it's kinda cute to commit
```tsx
// https://github.com/vercel/next.js/discussions/49415#discussioncomment-10388757
export const evilHtmlLangHack = {
	lang: "en"
};
```
to a repo.

But, `react-compiler` eslint is really unhappy with this.

**Reply by janvorwerk** [reply]
@hdodov my understanding is that this works... until two browser are hitting your page concurrently with different languages (but still in the same server instance). Not sure how much you tested it, but I would be scared that this solution is a _"it works on my machine"_ kind of trick.

**Reply by yandodov** [reply]
@janvorwerk what "server instance" are you talking about? This whole thing is when you're dealing with a **static website**. If you have a server, you won't have `output: "export"` and you'd use middleware like usual, without all these hacks.

My point was that my machine exports correctly the *static* website that *everyone* will see. I mean, it either works locally and for everyone or it doesn't work locally and for no one.

**Edit:** I saw that the OP's case wasn't about static sites in particular. So to clarify, the workaround I've given works well for static sites, i.e. with `output: "export"` only.

**Reply by richardtaylordawson** [reply]
This worked for my ISR/SSG version of a nextjs website, however, I did notice an issue if the issue doesn't use a NextLink to navigate the pages. If they go from an `en` to `de` page but then use back buttons in the browser it will stay as `de`

**Reply by vucasavovic** [reply]
Im not sure how can this work, since the root layout gets rendered/built first, than nested locale layout , than page that's changing the state ??

**Comment by poorscousertommy8** [comment]
We are facing the same issue: The content is loaded through a CMS, and the language of the content cannot be determined either by the domain or necessarily by the path. Of course, it’s clear that all content under /en/* is in English, but there are also short URLs where the language is unknown and can/must only be set after fetching the content. How can this issue be resolved?

**Reply by richardtaylordawson** [reply]
@poorscousertommy8 could you swap your urls/page pathing so that you can determine by the path? If so, I found a workaround for now via the middleware.js file you could do in order to get this to work

**Reply by richardtaylordawson** [reply]
@poorscousertommy8 Just noting since I have learned more on this issue 🤣 The middleware method only works with SSR. Essentially you look at the route, determine language, set the language in a cookie and use that cookie in your layout file. However, the issue then is it won't work with ISR/SSG because it uses cookies

**Comment by tranvanhieu01012002** [comment]
You can try this way:
Create a ```LangWrapper.jsx``` component


```
"use client";

import { usePathname } from "next/navigation";

const LangWrapper = ({ children }) => {
  const pathname = usePathname();
  //your locale in pathname
  const locale = pathname.split("/")[x];

  return <html lang={locale}>{children}</html>;
};

export default LangWrapper;
```
Use it in ```app/layout.jsx```
```
  return (
    <LangWrapper>
      <head>
         {/* Other head tags */}
      </head>
      <body>
       {children}
      </body>
    </LangWrapper>
```

**Reply by richardtaylordawson** [reply]
I believe that would change your entire app to client side rendering

**Reply by tranvanhieu01012002** [reply]
@richardtaylordawson 
no, you can keep the ```app/layout.jsx``` as ```"use server"```
It will be like your thinking, if we use a ssr component into a csr component

**Reply by ryami333** [reply]
You guys are missing the point a bit. The issue is that sometimes `lang` can't be trivially (or synchronously) derived from the pathname - it needs to be computed asynchronously (eg. fetched from a database). Yes, we can `await` that computation in `layout.tsx`, but often the dependent params are only available at the page-level, not the layout-level.

We basically have the exact same problem that the Metadata API solved for things like `<title>`. There are _parts_ of the layout that we might want to change on _every_ route, even while persisting other parts of the layout (eg. "Navbar"). The `lang` attribute (like `<title>`) is one of those things. There exists a mechanism to address this for `<title>` (the Metadata API), but not for `lang`.

**Reply by tranvanhieu01012002** [reply]
@ryami333  sorry, I understood your issue. 
You are right, the code above just works for simple ```lang``` attribute that exists in pathname.
With your case maybe we need to wait new updating from Nextjs core team.

**Reply by abolfazlakbarzadeh** [reply]
> You can try this way: Create a `LangWrapper.jsx` component
> 
> ```
> "use client";
> 
> import { usePathname } from "next/navigation";
> 
> const LangWrapper = ({ children }) => {
>   const pathname = usePathname();
>   //your locale in pathname
>   const locale = pathname.split("/")[x];
> 
>   return <html lang={locale}>{children}</html>;
> };
> 
> export default LangWrapper;
> ```
> 
> Use it in `app/layout.jsx`
> 
> ```
>   return (
>     <LangWrapper>
>       <head>
>          {/* Other head tags */}
>       </head>
>       <body>
>        {children}
>       </body>
>     </LangWrapper>
> ```

it can cause SEO issues, it would be  better to do these in server components, also it can cause hydration or/and SSR issues

**Comment by abolfazlakbarzadeh** [comment]
As an alternative, you can group your routes `(route-group-name)` and inside of each route group or in the first layout in route-group hierarchy, put the `<html lang="" dir="">` tag, in my case I needed to change `lang` and `dir` attributes based on business settings, so I put the `<html>` and `<body>` tags in a separate component and pass `lang` and `dir` to this component as props and I used this component in route groups.

let me show you by app router hierarchy:
```
app
-- layout.tsx remove html and body tags from this layout
-- (general route group) // put your page.tsx file here instead of keep it in app dir
---- layout.tsx // layout general in this layout I used html tag with static attrs
-- (business route group)
---- [business_uuid]
-------- layout.tsx // layout with dynamic html attrs, in this layout I used html tag with dynamic attrs
```

and it is working for me, of course I hope next.js support it by maybe metadata/generateMetadata or built-in components, but for now, It can work for us, I hope so :)

**Reply by eric-burel** [reply]
With this solution I fail to properly handle not-found routes, I still end up with "The following tags are missing in the Root Layout: <html>, <body>. Read more at https://nextjs.org/docs/messages/missing-root-layout-tags", and moving "not-found" in my "(root)" route group doesn't work. I am using Next 14.2 though maybe not-found works better with route groups in Next 15

**Reply by abolfazlakbarzadeh** [reply]
For this, I've defined not-found page file and wrapped it's content by needed tags

**Reply by jan-victor** [reply]
@leerob @rauchg  this i a  crucial issue. the api solution seems the most fitting. any solution or official recommendations in sight for that?

**Comment by yandodov** [comment]
I've found a workaround for this by using async layout and page, promises, and [server-only-context](https://www.npmjs.com/package/server-only-context) (which uses [React `cache()`](https://react.dev/reference/react/cache)).

`app/contexts.ts`:

```ts
import serverContext from "server-only-context";

export const [getLangResolve, setLangResolve] = serverContext<
  (lang: string) => void
>(() => {});
```

`app/layout.tsx`:

```tsx
import { setLangResolve } from "./contexts";

export default async function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  const lang = await new Promise<string>((resolve) => {
    setLangResolve(resolve);
    setTimeout(() => resolve("en"), 100); // fallback, just in case
  });
  return (
    <html lang={lang}>
      <body>{children}</body>
    </html>
  );
}
```

`app/[lang]/page.tsx`:

```tsx
import { getLangResolve } from "../contexts";

export default async function Home(props: {
  params: Promise<{ lang: string }>;
}) {
  const { lang } = await props.params;
  getLangResolve()(lang);
  return <h1 className="text-6xl">lang {lang}</h1>;
}
```

This basically does the following:

1. Layout starts rendering
2. Layout creates a promise for the lang code
3. Page starts rendering
   - **Note:** It looks like, **by default**, the page starts rendering *before* the layout has resolved and doesn't *need* it to be resolved. It's what makes this work.
5. Page resolves the lang promise using the param from its props
6. Layout resolves
7. Page does whatever async stuff it wants (make db calls, etc.)
8. Page resolves
9. Response returned

Generally, you can use this approach to pass whatever you want from the page up to the layout.

I tested it for race conditions and it's doing great. It has also been on production for a few days and I haven't seen any issues there as well.

Check a demo of it here: https://github.com/hdodov/test-nextjs/tree/demo-dynamic-lang

---

@leerob although this workaround seems to do the job, it'd be much better if Next offers a native solution that doesn't require such despicable promise gymnastics.

**Reply by aronwoost** [reply]
Compared to your previous `export const state = { locale: "en" };` approach, this solution has the advantage, that - because `React.cache()` us used - the `lang` value is **guaranteed** to be sticked to a user request. So the `lang` value can not be changed from another user requesting the page on the (almost exact) same time. Right?

What scares me a bit, is that it depends on the layout<->page rendering order. When Next.js internals change, this might break.

**Reply by aronwoost** [reply]
Source for the `React.cache()` behavior: https://react.dev/reference/react/cache#caveats

**Reply by yandodov** [reply]
@aronwoost true, but this solution requires middleware, which is not supported in SSG. The main benefit of the other solution is that it works with SSG and that was my main use case for it, actually.

**Reply by aronwoost** [reply]
I made your `server-only-context` solution work with SSG today. 😊

**Reply by aronwoost** [reply]
But I did some changes, now that I think of it. Maybe I reworked it, to be closer to the `export const state = { locale: "en" };` solution now. 😂😂😂

**Comment by leo-russi** [comment]
Hi, I solved it this way but I am not fully certain about the approach:

- I have a middleware forcing always a nextresponse url with a /[lang] 
- I have removed the RootLayout completely
- I have a LocaleRootLayout (the real root layout now) in src/app/layout.js which captures the params.lang and set the html lang attribute
- and I have a src/app/[lang]/[...notfound]/page.js to catch all wrong paths

Do you see any bad drawback by removing the RootLayout?

**Reply by yandodov** [reply]
The only downside I can think of is that you wouldn't get your custom layout for requests that **do not** get matched by your middleware. Static files, for example. In that situation, you'll get the default Next.js 404 page, I believe.

**Reply by leo-russi** [reply]
Not really. if someone is going to /wrong/path/to/img.png or /wrong/path/to/my.pdf the middleware takes over and then my [...notfound] catches and displays my custom PageNotFound correctly wrapped under my custom layout.

Hope next.js will soon make setting the html lang attribute a no brainer.

**Reply by yandodov** [reply]
Ah, so you're not using a matcher in your middleware and instead letting everything pass through it?

**Reply by leo-russi** [reply]
I am using a matcher. If the static resource is found all is ok. If not, it is going through the middleware which occurs in a not found path.
Here is my matcher:
'/((?!api|_next/static|_next/image|.*\.pdf|.*\.svg|favicon.ico|sitemap.xml|robots.txt).*)',

**Reply by leo-russi** [reply]
`'/((?!api|_next/static|_next/image|.*\.pdf|.*\.svg|favicon.ico|sitemap.xml|robots.txt).*)'`

**Reply by yunntan** [reply]
@leo-russi thanks for the cool solution, I found one drawback: you can't use [`notFound()`](https://nextjs.org/docs/app/api-reference/functions/not-found) anymore as it only redirects to the `not-found.js|ts` present at the `app` root level, which requires a root layout to work.

**Comment by julpat** [comment]
Hi, I have solved this in quite an easy way. It feels a bit hacky accessing `__NEXT_DATA__` but... 🤷‍♂️ 
`pageProps` are `props` passed in `getServerSideProps`. In my case, it's data fetched from CMS.


```
import Document from 'next/document'
import { Html, Head, Main, NextScript } from 'next/document'

class MyDocument extends Document {
	render() {
		const pageProps = this.props['__NEXT_DATA__'].props.pageProps

		return (
			// Language change triggers hard refresh so no client-side handling is implemented
			<Html lang={pageProps.locale.code}>
				<Head />
				<body>
					<Main />
					<NextScript />
				</body>
			</Html>
		)
	}
}

export default MyDocument
```

**Reply by julpat** [reply]
Sorry, this was on the Pages router.
Now I am trying to migrate to App Router to be able use cache, but got stuck on this issue. This is ridiculous.

**Comment by fikkatra** [comment]
For those of you, like me, don't mind using middleware: it seems Next.js official [documentation on internationalization](https://nextjs.org/docs/app/building-your-application/routing/internationalization#routing-overview) suggests an automatic redirect by the middleware:

```
import { NextResponse } from "next/server";
 
let locales = ['en-US', 'nl-NL', 'nl']
 
// Get the preferred locale, similar to the above or using a library
function getLocale(request) { ... }
 
export function middleware(request) {
  // Check if there is any supported locale in the pathname
  const { pathname } = request.nextUrl
  const pathnameHasLocale = locales.some(
    (locale) => pathname.startsWith(`/${locale}/`) || pathname === `/${locale}`
  )
 
  if (pathnameHasLocale) return
 
  // Redirect if there is no locale
  const locale = getLocale(request)
  request.nextUrl.pathname = `/${locale}${pathname}`
  // e.g. incoming request is /products
  // The new URL is now /en-US/products
  return NextResponse.redirect(request.nextUrl)
}
 
export const config = {
  matcher: [
    // Skip all internal paths (_next)
    '/((?!_next).*)',
    // Optional: only run on root (/) URL
    // '/'
  ],
}
```

They also mention:

> The root layout can also be nested in the new folder (e.g. app/[lang]/layout.js).

This is analogous to @leo-russi 's solution.

Having said that, it feels quite brittle to remove the root layout. I'd definitely prefer a much easier way to update the `lang` attribute of the `html` element.

**Comment by luizvergennes** [comment]
### ⚠️Transforms all your app into server-rendered on demand (Dynamic)!

Hey everyone, I'm a bit nervous to jump in, but I've finally got it working! I'm following Next.js's documentation approach to internationalization by embedding the locale in the URL slug. The key part of my implementation is in `middleware.ts`, where I use the following code:

```ts
// Within middleware.ts
const response = NextResponse.next();
response.headers.set("locale", locale);
return response;
```

This setup allows me to access the locale in my `rootLayout` like this:

```tsx
// Within rootLayout
const reqHeaders = await headers();
const locale = getLocale(reqHeaders);
return (
  <html lang={locale}>
    {/* ... */}
  </html>
);
```

**Reply by janvorwerk** [reply]
@Luk4h my understanding is that calling `headers()` turns a page to "dynamic" (SSG will be disabled on any page that uses this API). And since you are on the root layout, this would mean that your entire app would become dynamic.
Or did  I miss something?
BTW, this may or may not be an issue for you.

**Reply by luizvergennes** [reply]
You are correct!
Sadly now, all my pages are server-rendered on demand.
Not an issue to my specific use-case. But that is a great catch! Thanks! I will also update my comment to alert people who don't like comments 🙂.

**Comment by nghiaht** [comment]
Here is my solution:

Store the `lang` in the cookie `app-locale` and use it in `RootLayout` (`app/layout.tsx`).
No need to duplicate `<html lang>` in the `app/[lang]/layout.tsx`

```
// i18n-config.ts

export const i18n = {
  defaultLocale: "vi",
  locales: ["vi", "en"],
};

export type Locale = (typeof i18n)["locales"][number];
```

```
// middleware.ts

export function middleware(request: NextRequest) {
  // Check if there is any supported locale in the pathname
  const { pathname } = request.nextUrl;

  let foundLocale: Locale | undefined;

  for (const locale of i18n.locales) {
    if (pathname.startsWith(`/${locale}/`) || pathname === `/${locale}`) {
      foundLocale = locale;
      break;
    }
  }

  if (foundLocale) {
    const response = NextResponse.next({});

    const cookieLocale = request.cookies.get("app-locale")?.value;
    if (cookieLocale !== foundLocale) {
      response.cookies.set("app-locale", foundLocale);
    }

    return response;
  }

  // Redirect if there is no locale
  const locale = getLocale(request);
  request.nextUrl.pathname = `/${locale}${pathname}`;

  // e.g. incoming request is /products
  // The new URL is now /en-US/products
  const response = NextResponse.redirect(request.nextUrl);

  response.cookies.set("app-locale", locale!);

  return response;
}
```

```
// app/layout.tsx


import { cookies } from "next/headers";

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {

  const theCookies = await cookies()
  const locale = theCookies.get("app-locale")?.value;


  return (
    <html lang={locale}>
      <body>{children}</body>
    </html>
  );
}

```

**Comment by janvorwerk** [comment]
Not sure if I missed something from the initial requirement, but I have been using following pattern for a while w/o any trouble...

1. I have a middleware that systematically redirects to a specific `[lang]` path (ex: `/en/foo`, `/fr/foo`, ...)
2. The root layout is a simple pass-through (see below)
3. Another layout in the `[lang]/` folder contains the `<html lang=` based on the parameter

Pass-through root layout in `app/` folder

```tsx
export default function RootLayout(props: { children: React.ReactNode }) {
  return props.children;
}
```

Having this kind of root layout makes the 404 page work.

Hope this helps, and kindly let me know if I missed something.



--edit : my `app/not-found.tsx` page has to render the `<html>` tag (since it's not present in the `app/layout.tsx`. Here I use a lang cookie set by the middleware to set the lang attribute.

**Reply by nghiaht** [reply]
Brilliant.
I didn't think that return `props.children` in RootLayout which works!

**Comment by jasonday** [comment]
Followed this approach: https://www.meje.dev/blog/html-lang-attribute-in-nextjs

Someone enlighten me if this is terrible. 

> Alternatively, you can fix this issue by going into the pages route and create a file, let's call it _document.js, and proceed to add the content of the snippet below into it.

```
import { Html, Head, Main, NextScript } from 'next/document'
 
export default function Document() {
  return (
    <Html lang="en">
      <Head />
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  )
}
```

**Reply by MonstraG** [reply]
That's for pages router, this discussion talks about app router.

**Comment by phanminhtai** [comment]
It's almost April 2025, and this question that should have been addressed earlier still doesn't have an answer :(

**Reply by nghiaht** [reply]
Did you try this: https://github.com/vercel/next.js/discussions/49415#discussioncomment-12300860?

**Comment by ryami333** [comment]
Not without compromise, no. The issue and subsequent clarifying comments here have been very deliberately worded to demonstrate that this issue must be fixed upstream, because otherwise you subvert the concept of persisted shared layouts entirely. This needs an API for the _exact_ same reason that Metadata needed its own API (because metadata exists so high in the DOM hierarchy that to change it between individual routes would invalidate layouts).

**Reply by janvorwerk** [reply]
Not sure what is the compromise? Except if your users keep changing language all the time, the persisted shared layouts work… for each locale individually. To me Metadata are another story because each page has its own title, etc…

**Reply by janvorwerk** [reply]
@ryami333: adding a thumb down does not really answer my question: _"what is the compromise?"_.
My [workaround](https://github.com/vercel/next.js/discussions/49415#discussioncomment-12300860), added to [@luizvergennes 's](https://github.com/vercel/next.js/discussions/49415#discussioncomment-12557922) may not suit your needs but it may work for most users finding this issue…

**Comment by luizvergennes** [comment]
Some other alternative, since we have yet to find the ideal one... This one needs to follow the `[lang]` slug structure as the Next.js docs. In your layout, you can create a function `generateStaticParams` that "mocks" expected params. With this approach, you are able to build the pages as SSG. (ps. Take a look @janvorwerk! ) 

```typescript
export async function generateStaticParams() {
    return [{ lang: "en-US" }, { lang: "pt-BR" }];
}

interface DashboardLayoutProps {
    children: React.ReactNode;
    params: Promise<{ lang: Locale }>;
}

export default async function DashboardLayout({
    params,
    children,
}: DashboardLayoutProps) {
    const { lang } = await params;
    ...
```

Result of running `next build` on my project with this:
```sh
Route (app)                              Size     First Load JS
┌ ○ /_not-found                          979 B           106 kB
├ ● /[lang]/dashboard                    3.77 kB         149 kB
├   ├ /en-US/dashboard
├   └ /pt-BR/dashboard
├ ● /[lang]/forgot-password              2.92 kB         187 kB
├   ├ /en-US/forgot-password
├   └ /pt-BR/forgot-password
├ ● /[lang]/onboarding                   1.1 kB          319 kB
├   ├ /en-US/onboarding
├   └ /pt-BR/onboarding
├ ● /[lang]/pricing                      2.28 kB         114 kB
├   ├ /en-US/pricing
├   └ /pt-BR/pricing
├ ● /[lang]/referrals                    9.05 kB         221 kB
├   ├ /en-US/referrals
├   └ /pt-BR/referrals
├ ● /[lang]/reset-password               4 kB            188 kB
├   ├ /en-US/reset-password
├   └ /pt-BR/reset-password
├ ● /[lang]/services                     7.86 kB         186 kB
├   ├ /en-US/services
├   └ /pt-BR/services
├ ● /[lang]/signin                       4.92 kB         191 kB
├   ├ /en-US/signin
├   └ /pt-BR/signin
├ ● /[lang]/signup                       3.23 kB         319 kB
├   ├ /en-US/signup
├   └ /pt-BR/signup
├ ● /[lang]/verify                       6.89 kB         188 kB
├   ├ /en-US/verify
├   └ /pt-BR/verify
├ ƒ /api/[...nextauth]                   144 B           105 kB
├ ƒ /api/stripe/checkout                 144 B           105 kB
├ ƒ /api/stripe/webhook                  144 B           105 kB
└ ƒ /api/uploadthing                     144 B           105 kB
+ First Load JS shared by all            105 kB
  ├ chunks/4bd1b696-a9b262982930989a.js  52.9 kB
  ├ chunks/517-f39a7a794abbf6cb.js       50.5 kB
  └ other shared chunks (total)          1.92 kB


ƒ Middleware                             110 kB

○  (Static)   prerendered as static content
●  (SSG)      prerendered as static HTML (uses generateStaticParams)
ƒ  (Dynamic)  server-rendered on demand
```

edit: grammar

**Reply by MonstraG** [reply]
With this approach, 
1) do you have a root layout?
2) where are the `<body>` and `<html>` tags placed?
3) did you set correct `lang` tag on the body?
4) does the `not-found` page work out of the box?

I have found that in order to do `3` I have to completely avoid using `not-found` and do my own `[...fallback]/page.tsx` (that is actually not found)

**Reply by luizvergennes** [reply]
With the exception of item 4, most of the solutions work. I did not encounter an issue with the not-found page because I am primarily working on it in a dashboard project. In this project, my middleware intercepts all requests that are not already mapped to a constant object and redirects them to the home dashboard. I have yet to find the time to further develop this solution and achieve your desired outcome.
Recently, [Vercel announced that they have finally implemented the ability to send the metadata of a page after the initial load](https://www.linkedin.com/posts/delbaoliveira_nextjs-152-streaming-metadata-async-ugcPost-7301502032857677824--6Pe?utm_source=share&utm_medium=member_desktop&rcm=ACoAAC4nfqUBKLdGskbJQDL5Q0yJvC34V2jt_AU). This development indicates that we are approaching an official solution to this problem.

**Comment by richardtaylordawson** [comment]
@leerob, Hoping you can either shed some light on this issue or point us in the right direction of who can help. This issue is a massive one and I'm shocked it's not a bigger issue at hand considering the workarounds aren't full proof and the only one that seems to guarantee the SSG side of things is to have a client component update the lang (which is bad for SEO which is typically the reason for SSG in the first place)

Let me know, would love to be able to find a solution!

cc: @ryami333 @ViliamKopecky tagging you both as well in case you've found anything since the start of this convo. I've let the workaround set in my codebase awhile now, but the upgrade to 15/16 seem to be making the workaround not be a workaround and a bigger issue now 😆
